File: 110-balanced-binary-tree/balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Pair{
    public:
        int ht;
        bool isbal;
};
class Solution {
public:
    Pair* check(TreeNode* root){
        if(root == NULL){
            Pair* bp = new Pair;
            bp->ht = 0;
            bp->isbal = true;
            return bp;
        }
        Pair* left = check(root->left);
        Pair* right = check(root->right);

        Pair* mp = new Pair();
        mp->isbal = abs(left->ht - right->ht) <= 1 
                    && left->isbal && right->isbal;
        mp->ht = max(left->ht, right->ht) + 1;
        return mp;
    }

    bool isBalanced(TreeNode* root) {
        Pair* mp = check(root);
        return mp->isbal;
    }
};

File: 110-balanced-binary-tree/balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Pair{
        int ht;
        boolean isBal;
        Pair(int ht, boolean isBal){
            this.ht = ht;
            this.isBal = isBal;
        }
        Pair(){}
    }

    private Pair checkBalance(TreeNode root){
        if(root == null){
            Pair bp = new Pair(0, true);
            return bp;
        }
        Pair left = checkBalance(root.left);
        Pair right = checkBalance(root.right);

        Pair mp = new Pair();
        
        mp.isBal = Math.abs(left.ht - right.ht) <= 1
                    && left.isBal && right.isBal;
        mp.ht = Math.max(left.ht, right.ht) + 1;
        return mp;
    }

    public boolean isBalanced(TreeNode root) {
        // Code here
        Pair ans = checkBalance(root);
        return ans.isBal;
    }
}